src/neo/SmartContract/ApplicationEngine.Binary.cs

```diff
@@ -23,6 +28,26 @@ protected internal StackItem BinaryDeserialize(byte[] data)
             return BinarySerializer.Deserialize(data, Limits.MaxStackSize, Limits.MaxItemSize, ReferenceCounter);
         }
 
+        protected internal string Itoa(BigInteger value, int @base)
+        {
+            return @base switch
+            {
+                10 => value.ToString(),
+                16 => value.ToString("x"),
+                _ => throw new ArgumentOutOfRangeException(nameof(@base))
+            };
+        }
+
+        protected internal BigInteger Atoi(string value, int @base)
+        {
+            return @base switch
+            {
+                10 => BigInteger.Parse(value),
+                16 => BigInteger.Parse(value, NumberStyles.HexNumber, CultureInfo.InvariantCulture),
```

Comment:
```suggestion
                16 => BigInteger.Parse(value, NumberStyles.AllowHexSpecifier, CultureInfo.InvariantCulture),
```

We should not allow leading or trailing white-space characters.